system/jlib/jdebug.hpp

```diff
@@ -225,6 +226,108 @@ class jlib_decl HiresTimer
+ * When an items is dequeued/complete atomically decrement a counter and add the current timestamp to the accumulated end time stamps.
+ * To calculate the waiting time use (sumEndTimestamps + numWaiting * currentTimeStamp) - sumStartTimestamps.
+
+
+At first glance this appears to have a problem because sumEndTimestamps will quickly overflow, but because of the
+properties of modulo arithmetic, the final result will be correct even if it has overflowed!
+Also since you are only ever interested in (sumEndTimestamps - sumStartTimestamps) the same field can be used for both.
+
+There are two versions, one that uses a critical section, and a second that uses atomics, but is limited to the number
+of active blocked items.
+*/
+
+class jlib_decl BlockedTimeTracker
+{
+public:
+    BlockedTimeTracker() = default;
+    BlockedTimeTracker(const BlockedTimeTracker &) = delete;
+
+    void noteWaiting();
+    void noteComplete();
+    __uint64 getWaitingNs() const;
+
+private:
+    mutable CriticalSection cs;
+    unsigned numWaiting = 0;
+    cycle_t timeStampTally = 0;
+};
+
+class jlib_decl BlockedSection
+{
+public:
+    BlockedSection(BlockedTimeTracker & _tracker) : tracker(_tracker) { tracker.noteWaiting(); }
+    ~BlockedSection() { tracker.noteComplete(); }
+private:
+    BlockedTimeTracker & tracker;
+};
+
+//Lightweight version that uses a single atomic, but has a limit on the number of active blocked items
+//Easier to understand by looking at the code for the BlockedTimeTracker class
+//Could be a template class (for MAX_ACTIVE), but I doubt it is worth it.
+class jlib_decl LightweightBlockedTimeTracker
+{
+    //MAX_ACTIVE should be a power of 2 for efficiency, 256 gives a max blocked time of about half a year before wrapping.
+    static constexpr unsigned MAX_ACTIVE = 256;
+public:
+    LightweightBlockedTimeTracker() = default;
+    LightweightBlockedTimeTracker(const LightweightBlockedTimeTracker &) = delete;
+
+    void noteWaiting()                  { timeStampTally.fetch_sub((get_cycles_now() * MAX_ACTIVE) - 1); } // i.e. add 1 and subtract the time
+    void noteComplete()                 { timeStampTally.fetch_add((get_cycles_now() * MAX_ACTIVE) - 1); }
+    __uint64 getWaitingNs() const;
+
+private:
+    std::atomic<__uint64> timeStampTally{0};            // timestamp * MAX_ACTIVE + active
+};
+
+class jlib_decl LightweightBlockedSection
+{
+public:
+    LightweightBlockedSection(LightweightBlockedTimeTracker & _tracker) : tracker(_tracker) { tracker.noteWaiting(); }
+    ~LightweightBlockedSection() { tracker.noteComplete(); }
+private:
+    LightweightBlockedTimeTracker & tracker;
+};
+
+
+// Utility class and function for performing an action when the object is destroyed.
+// Example of use:
+// auto action = OnLeave([&](){ timer.noteComplete(); });
+template <class DESTRUCTOR>
+class OnLeaveAction
```

Comment:
This is very similar to the existing COnScopeExit.

Reply:
Yes, that is doing the same - I will remove mine.  How do we make it clear where classes like that can be found.